tests/unit/test_cloudwatch.py

```diff
@@ -0,0 +1,228 @@
+        |TEST_2_M_OF_N | 0 0 X 0 X    | ALARM            | ALARM       | ALARM     | ALARM         | 0                    |
+        |TEST_3_M_OF_N | 0 - X - -    | OK               | OK          | ALARM     | OK            | 1                    |
+        |TEST_4_M_OF_N | - - - - 0    | OK               | OK          | ALARM     | OK            | 2                    |
+        |TEST_5_M_OF_N | - - - X -    | ALARM            | Retain state| ALARM     | OK            | 2 -> Premature alarm |
+        """
+
+        def mock_metric_alarm_details(alarm_arn):
+            details = {
+                "AlarmName": "test-alarm",
+                "StateValue": initial_state,
+                "TreatMissingData": treat_missing,
+                "DatapointsToAlarm": 2,
+                "EvaluationPeriods": 3,
+                "Period": 5,
+                "ComparisonOperator": "LessThanThreshold",
+                "Threshold": 2,
+            }
+            return details
+
+        def mock_collect_metric_data(alarm_details, client):
+            if metric_data == self.TEST_1_M_OF_N:
+                return [2.5, None, 1, None, 1.7]
+            if metric_data == self.TEST_2_M_OF_N:
+                return [3.0, 4.0, 1.0, 3.5, 1.5]
+            if metric_data == self.TEST_3_M_OF_N:
+                return [4.0, None, 1.2, None, None]
+            if metric_data == self.TEST_4_M_OF_N:
+                return [None, None, None, None, 8]
+            if metric_data == self.TEST_5_M_OF_N:
+                return [None, None, None, 1.0, None]
+
+        run_and_assert_calculate_alarm_state(
+            mock_metric_alarm_details, mock_collect_metric_data, expected_calls, expected_state
+        )
+
+
+def run_and_assert_calculate_alarm_state(
+    mock_metric_alarm_details, mock_collect_metric_data, expected_calls, expected_state
+):
+    mock_client = Mock()
+
+    def mock_cloudwatch_client(alarm_arn):
+        return mock_client
+
+    alarm_scheduler_pckg = "localstack.services.cloudwatch.alarm_scheduler"
+    with patch(
+        f"{alarm_scheduler_pckg}.get_metric_alarm_details_for_alarm_arn",
+        mock_metric_alarm_details,
+    ):
+        with patch(
+            f"{alarm_scheduler_pckg}.get_cloudwatch_client_for_region_of_alarm",
+            mock_cloudwatch_client,
+        ):
+            with patch(
+                f"{alarm_scheduler_pckg}.collect_metric_data",
+                mock_collect_metric_data,
+            ):
+                calculate_alarm_state("helloworld")
+                assert len(mock_client.mock_calls) == expected_calls
+                if expected_calls != 0:
+                    expected_calls = [
+                        call.set_alarm_state(
+                            AlarmName="test-alarm", StateValue=expected_state, StateReason=ANY
+                        )
+                    ]
+                    mock_client.assert_has_calls(expected_calls)
```

Comment:
this test could be cleaned up by using the `monkeypatch` fixture that [pytest provides](https://docs.pytest.org/en/6.2.x/monkeypatch.html#simple-example-monkeypatching-functions)

Reply:
i cleaned it up with our own patch utility :-)